### PR TITLE
Added AcceleratedOptimizer

### PR DESCRIPTION
**Pull request by degleris1**
Two major changes
- Added AcceleratedOptimizer and `A-HALS` algorithm. A similar `A-MU` algorithm won't be too hard to create.
- Moved `EPSILON` (for division) to `common.py`, so all modules use the same `EPSILON`.

### CODE REVIEW

File: cmfpy/algs/accelerated.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    def __init__(self, data, dims, patience=3, tol=1-5,
+                 max_inner=0, inner_thresh=0, weightW=1, weightH=1,
+                 **kwargs):
+        super().__init__(data, dims, patience=patience, tol=tol, **kwargs)
+
+        self.max_inner = max_inner
+        self.inner_thresh = inner_thresh
+        self.weightW = weightW
+        self.weightH = weightH
```
Comment (ahwillia) [inline cmfpy/algs/accelerated.py]: Can you define what all of these parameters (max_iter, inner_thres, weightW, weightH) mean in a docstring for `__init__`?

File: cmfpy/algs/accelerated.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        H_hist.append(self.H.copy())
+        H_init_change = la.norm(H_hist[-1] - H_hist[-2])
+
+        # Update several more times
+        for itr in range(int(self.max_inner * self.weightH)):
+            self.update_H()
+            H_hist.append(self.H.copy())
+            if (self.inner_thresh > 0 and la.norm(H_hist[-1] - H_hist[-2])
+                    < self.inner_thresh * H_init_change):
+                break
```
Comment (ahwillia) [inline cmfpy/algs/accelerated.py]: This code chunk is replicated twice for updating `W` and `H`. Consider refactoring it into a function.

```python

def update(self):

    ...

    # Update W
    self.setup_W_update()
    self._accelerated_update(self.W, self.update_W)

    # Update H
    self.setup_H_update()
    self._accelerated_update(self.H, self.update_H)

    ...

def _accelerated_update(self, param, update_func):

    while itr < self.max_inner:
        update_func()
        param_hist.append(param.copy())
```

File: cmfpy/algs/hals.py [hunk 3]
```diff
@@ -85,73 +91,98 @@ def next_H_entry(self, Wkt, norm_Wkt, remainder):
         """
         trace = np.dot(np.ravel(Wkt), np.ravel(-remainder))
 
-        return np.maximum(trace / (norm_Wkt**2 + EPSILON), 0)
+        return np.maximum(trace / (norm_Wkt**2 + EPSILON), FACTOR_MIN)
+
+
+"""
+"""
```
Comment (ahwillia) [inline cmfpy/algs/hals.py]: Empty comment here? Delete?

File: cmfpy/algs/hals.py [hunk 4]
```diff
@@ -1,80 +1,86 @@
...
 import numpy.linalg as la
 
-from .base import AbstractOptimizer, EPSILON
-from ..common import shift_and_stack, cmf_predict
+from .accelerated import AcceleratedOptimizer
+from ..common import shift_and_stack, EPSILON, FACTOR_MIN
 
 
-class SimpleHALSUpdate(AbstractOptimizer):
+class SimpleHALSUpdate(AcceleratedOptimizer):
```
Comment (ahwillia) [inline cmfpy/algs/hals.py]: I feel like we should call this `AcceleratedHALS` or something... Or at least make `max_inner = 0` by default.

Comment (degleris1) [reply cmfpy/algs/hals.py]: I've set the default to a single iteration (i.e. regular HALS).

File: cmfpy/algs/hals.py [hunk 5]
```diff
@@ -1,181 +1,217 @@
...
-        """
-        return np.outer(np.ones(n_batch), Wk)
+
+@numba.jit(nopython=True)
+def _fold_factor(Wk, batch):
+    """
+    Generate factor prediction for a given component and lag. Then fold
+    into a tensor.
+    """
+    return np.outer(batch, Wk)
```
Comment (ahwillia) [inline cmfpy/algs/hals.py]: No reason numba should give a speedup here.

**Pull request merged**